src/backend/utils/time/snapmgr.c

```
@@ -299,6 +299,105 @@ SnapMgrInit(void)
 	}
 }
 
+/*
+ * GPDB: get the snapshot based on restore point settings.
+ *
+ * Upon successful retrival of a restore point based snapshot, set CurrentSnapshot
+ * to that snapshot and return true.
+ */
+static bool GetRestorePointBasedSnapshot(void)
+{
+	RestorePointInfo 		rp;
+	char 				rpname[MAXFNAMELEN];
+	char 				snapshotname[MAXFNAMELEN + sizeof(RP_SNAPSHOT_PREFIX)];
+	bool 				found;
+
+	/* Irrelevant if not using restore point based snapshot. */
+	if (gp_hot_standby_snapshot_mode != HS_SNAPSHOT_RESTOREPOINT)
+		return false;
+
+	/* Irrelevant for primary nodes. */
+	if (!RecoveryInProgress())
+		return false;	
+
+	/* Irrelevant for single-node mode, such as utility or mainenance mode (mostly just for tests). */
+	if (Gp_role == GP_ROLE_UTILITY || gp_maintenance_mode)
+		return false;	
+
+	/* Skip if we are creating fresh snapshot when replaying RP. */
+	if (DistributedTransactionContext == DTX_CONTEXT_REPLAY_RP)
+		return false;
+
+	/*
+	 * We will be using restore point based snapshot, but skip if we 
+	 * already set one in the current transaction.
+	 */
+	if (FirstSnapshotSet)
+		return true;
+
+	/*
+	 * Now get the restore point name. 
+	 * For QD, it is either specified by GUC or just using the latest restore point.
+	 * For QE, it is passed via distributed snapshot dispatched from QD.
+	 */
+	memset(rpname, 0, MAXFNAMELEN);
+	if (IS_HOT_STANDBY_QD())
+	{
+		if (!gp_hot_standby_snapshot_restore_point_name)
+		{
+			ereport(DEBUG1,
```

What is the guarantee that shmLatestRpName is set here?

I don't think we should go into the business of using the latest snapshot if the GUC is not set. We want this to be very explicit.

I think we should throw a FATAL error here if `gp_hot_standby_snapshot_restore_point_name` is not set but `gp_hot_standby_snapshot_mode == HS_SNAPSHOT_RESTOREPOINT`.

This can be similar to:
```
			ereport(FATAL,
					(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
					 errmsg("database \"%s\" is not currently accepting connections",
							name)));
```

We should be doing this check much earlier, in `InitPostgres()`, where we should FATAL out. In this function, we should just Assert that the GUC's value is set to something sane.

---

When `shmLatestRpName` is NULL, we would error out later at `cannot find snapshot to use for restore point`.

I think eventually we should provide the option to use the latest RP, either by setting to `NULL` or something more explicit like `latest`. It would be very common for a user in wanting to use whatever the latest they can use. The orchestration between primary and standby would be too annoying. However, I do agree that at this point, allowing latest isn't too helpful, because (1) DR is going to always set it; (2) w/o DR, we don't have a way to make sure every segment has replayed the latest RP, so query can error out easily if one segment lags behind. We should use the *last-universally-replayed* RP instead of the *last-QD-replayed* RP. So I've removed that shmLatestRpName business.